Synchronous service create and service update

aaronlehmann opened the pull request:
Change "service create" and "service update" to wait until the creation or update finishes, when `--detach=false` is specified. Show progress bars for the overall operation and for each individual task (when there are a small enough number of tasks), unless `-q` / `--quiet` is specified.

Internals:

- [ ] The current implementation is polling. It would be much better to use the Swarm events API once it's ready (@dongluochen).
- [x] The task endpoint spec is not included when evaluating if a task is up to date. This structure isn't exposed through the REST API, so it would need to be added (or we need to come up with a better way of checking if a task is up to date).
- [x] The method of checking for global service convergence is a terrible hack. The constraint evaluator was ported to the client so it can tell which nodes match the constraints. This needs to be redone in a better way.
- [x] Handle rollback update states (once #31108 is merged)

Cosmetic/UX:

- [x] The progress bars show numbers like "9B/9B", assuming the units are bytes. This should be changed to drop the "B" suffix.
- [x] Upon `^C`, the command should print a notice that the creation or update is still continuing in the background.
- [x] Possible visual improvements: line up progress bars, etc.

cc @docker/core-swarmkit-maintainers @tiborvass @anusha-ragunathan

Commit:
Synchronous service create and service update

Change "service create" and "service update" to wait until the creation
or update finishes, when --detach=false is specified. Show progress bars
for the overall operation and for each individual task (when there are a
small enough number of tasks), unless "-q" / "--quiet" is specified.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

aaronlehmann commented:
Demo: https://asciinema.org/a/4btgvuk1dv3z8zn0y882gqph5

vdemeester commented:
@aaronlehmann what happens if the service created can't be scheduled, because of constraint that cannot be fullfilled for example ? How should it behave ?

thaJeztah commented:
I like the feature but it would be a huge breaking change. It feels a bit like the `docker-compose up` vs `docker-compose up -d` debate.

I can see use-cases for this, but should we also attach to the service logs after the service was deployed (like `docker-compose up`)?

Let me add this to the next maintainers meeting to get more eyes / open the discussion

aaronlehmann commented:
> what happens if the service created can't be scheduled, because of constraint that cannot be fullfilled for example ? How should it behave ?

You would see one or more of the tasks stuck at "pending". This synchronous mode gives visibility into the problem, which would otherwise require a "service ps" command to notice.

thaJeztah commented:
> You would see one or more of the tasks stuck at "pending".

Would it output the error messages  / reason for being unable to deploy the task?

aaronlehmann commented:
Not at present, but that's definitely possible.

dnephin commented:
+1 for the feature, -1 for changing the default immediately.

I think we need to put it behind a flag for now. We can add a warning when no flag is used, and change the default in a release or two.

Changing the default is going to break people who use this command in scripts.

thaJeztah commented:
We should also look at this for `docker stack deploy`, to show progress, and ideally show logs as well for a stack (to be more like `docker-compose`)

We discussed this in the maintainers meeting, and like the feature, but (for now) to not be the default.

aluzzardi commented:
Huge +1

It's also going to "fix" the workflow for many people that currently do a `service create` then just a bunch of `service ps` and then have to inspect tasks to figure out what happened.

Or worse - people that just do `service create`. Contrary to `docker run`, a service is not immediately ready and service creation is fully asynchronous. So the service is not responding, the port doesn't work, and the rollout may have failed.

UX Feedback:
- Everything should be aligned. I think the state should be padded to the longest possible value so the progress bars don't "shift" to the left/right when the state changes
- What's the `XB/XB` at the end? 
- `Waiting to verify that tasks are stable` - this is a bit non obvious, some rewording might help. Also, could we have a countdown? e.g. `Waiting X seconds`
- Not sure if `overall` needs its own progress bar. By definition it has a slightly different meaning than the others. On first look it seems like a slow task. Maybe plan text would be better, e.g. `Updating service <foo> (2 out of 12 completed)`
- What's interesting is error handling here. Could you link to a demo? I think the interesting bits of information are the overall status (the one in `UpdateStatus` - which I think should be part of the first line), the task error (can we show it somewhere?) and the rollback/pause status
- When the user Ctrl-C's the deployment, we should say that it's continuing in the background
- When we rollback, we should print the error message etc
- When we pause, we should say why and how to continue

Looks amazing so far!

diogomonica commented:
Looks awesome.

- Should Ctrl-C rollback the deployment?
- I think this going on by default might be too much, but an `--interactive` flag sounds good.

icecrime commented:
I'm with @dnephin: huge 👍, but I wouldn't change the default.

anusha-ragunathan commented:
Nice work!

Regarding re-usability of the service progressbar for other clients that potentially want to display service updates (eg, `plugin install`), it is better to extract `cli/command/service/progress.go` into a separate package.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "synchronous-service-commands" git@github.com:aaronlehmann/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842355079112
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

aaronlehmann commented:
Thanks for all the comments. I've updated this to address the feedback. Details below, and new demo here: https://asciinema.org/a/9f9rgk0e7fczkxi5vfu7yofmw

> I think we need to put it behind a flag for now. We can add a warning when no flag is used, and change the default in a release or two.

I've changed the behavior so the default is unchanged, and the `-i` / `--interactive` flag is required to see progress info. One thought I had is that if we only enable this by default when the command is running on a TTY, it's less likely we would break scripts. But I'm also okay with putting this behind an option, for now. It also means there's less urgency in fixing the technical implementation issues described in the PR description (though I would like to fix at least some of them before merging).

> Everything should be aligned. I think the state should be padded to the longest possible value so the progress bars don't "shift" to the left/right when the state changes

Done.

> What's the XB/XB at the end?

This was an artifact of the progress bar code. I've now added an option that suppresses it.

> Waiting to verify that tasks are stable - this is a bit non obvious, some rewording might help. Also, could we have a countdown? e.g. Waiting X seconds

Changed this so it has a countdown now.

> Not sure if overall needs its own progress bar. By definition it has a slightly different meaning than the others. On first look it seems like a slow task. Maybe plan text would be better, e.g. Updating service <foo> (2 out of 12 completed)

Changed to plain text

> What's interesting is error handling here. Could you link to a demo? I think the interesting bits of information are the overall status (the one in UpdateStatus - which I think should be part of the first line), the task error (can we show it somewhere?) and the rollback/pause status

Demo linked above. I've done some more testing and refinement of this now. Unfortunately it's not especially straightforward to show the task error, because the ID of the task that paused an update is inside a textual message. But we can figure something out if this would be a big usability win.

> When the user Ctrl-C's the deployment, we should say that it's continuing in the background

Done

> When we rollback, we should print the error message etc

Blocked on #31108

> When we pause, we should say why and how to continue

It says why. I'm not sure about the "how to continue" part. Any suggestions for what the message should say?

> Regarding re-usability of the service progressbar for other clients that potentially want to display service updates (eg, `plugin install`), it is better to extract `cli/command/service/progress.go` into a separate package.

Moved to `github.com/docker/docker/cli/command/service/progress`.

aluzzardi commented:
> It says why. I'm not sure about the "how to continue" part. Any suggestions for what the message should say?

`docker service update <foo>` should continue the paused deployment right?

Regarding the verification time - how are we picking the 45 seconds?

Also, is this looking at task status or update status? As in, do we take into account the rollback threshold?

aaronlehmann commented:
> `docker service update <foo>` should continue the paused deployment right?

Yeah, but I'm not sure we want to recommend that, because usually there's an underlying problem that should be fixed.

> Regarding the verification time - how are we picking the 45 seconds?

I set `--update-monitor 45s` when I created the service, to demonstrate failures (by manually stopping some containers). The default is `5s`.

> Also, is this looking at task status or update status? As in, do we take into account the rollback threshold?

Sort of a mix of both. Basically, if the update pauses or rolls back, we'll see that in the update status. But the progress bars and countdown are driven by task status.

aluzzardi commented:
Looks great :)

Do we have some CLI designers in the house? @dnephin feedback?

dnephin commented:
Design LGTM

WTFKr0 commented:
Hi
Is it possible to manage "docker service rm" action too ?
And what about "docker stack deploy/rm" command which launch multiple "docker service" actions ?
Thanx for reading

Commit:
Add support for UpToDate filter, for internal use

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "synchronous-service-commands" git@github.com:aaronlehmann/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842355076536
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

aaronlehmann commented:
I've updated this to check which tasks are up-to-date on the manager side (using a task filter), rather than using the incomplete check I had implemented in the client which couldn't consider `Endpoint`. This checks one item off the to-do list.

The only remaining thing that I think should block merging this is finding a better way to enumerate relevant nodes for global services. I have a similar idea for how to handle this.

aaronlehmann commented:
I found a way to avoid evaluating constraints in the client. Instead of computing how many nodes are eligible to run the service, it now takes the set of nodes seen referenced by up-to-date tasks. This is a reasonable proxy for computing that set directly.

I think the major issues have been resolved, and this is ready for code review and possibly merging. Avoiding polling would be nice, but that won't be possible for awhile. I think having an initial version that relies on polling is acceptable, particularly since this feature is activated by a flag rather than being the default.

aaronlehmann commented:
Test failure should be resolved by #31535

aaronlehmann commented:
Rebased and updated to report progress of rollback operations.

Here's an example of what the progress output looks like from a failed update that gets rolled back:

```
overall progress: rolling back update: 5 out of 5 tasks
1/5: running   [==================================================>]
2/5: running   [==================================================>]
3/5: running   [==================================================>]
4/5: running   [==================================================>]
5/5: running   [==================================================>]
rollback: update rolled back due to failure or early termination of task snr2t11e0t2t6lyzd6ule07rq
verify: Waiting 1 seconds to verify that tasks are stable...
```

PTAL. Should be ready to go unless there are suggestions for tweaks.

aaronlehmann commented:
Rebased

Review ping

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+				longest = len(state)
+			}
+		}
+		return longest
```
Minor: I usually see this kind of code in a `func init()` instead of `var`. I'm not sure if there is any real difference or if it's just best practice.

aaronlehmann replied:
Moved to an `init` function

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+			} else {
+				return errors.New("unrecognized service mode")
+			}
+		}
```
This block could be a new function

```go
updated, err = initializeUpdater(service, progressOut)
```

aaronlehmann replied:
Done

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+				if u.done {
+					if task.Status.State != swarm.TaskStateRunning {
+						u.done = false
+						goto progressBars
```
Could this be a function call instead of a goto?

aaronlehmann replied:
I tried this and ended up with the following function signature:

```
func (u *replicatedProgressUpdater) progressBars(activeNodes map[string]swarm.Node, tasksBySlot map[int]swarm.Task, replicas uint64) uint64 {
```

It returns `running`, which is computed inside the loop. The function recurses in the case where there's currently a goto.

TBH I find the version with the goto very slightly simpler after going through the exercise. But I'd be fine with switching to the function.

dnephin replied:
How about an inline function?

aaronlehmann replied:
A closure that calls itself is a little weird. It involves unusual syntax to first define the function as a variable, then assign it a function body. What's wrong with a goto?

dnephin replied:
It does unfortunately require some unusual syntax in go, but otherwise is a pretty common way of doing this.

I think the behaviour of a closure is much better understood.

aaronlehmann replied:
@dnephin: I've replaced this with code that doesn't involve a goto or recursion. PTAL.

dnephin replied:
Thanks, I like it

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+			}
+		}
+
+		converged, err = updater.update(ctx, service, tasks, activeNodes, rollback)
```
It looks like `activeNodes` is only used by the updater, and the only requirement is `ctx`, which is already being passed to `update()`. Could this be a `getActiveNodes(ctx)` function called from the updater, and the `activeNodes` parameter can be removed from the interface?

aaronlehmann replied:
As above, this would require the updater to have the `APIClient`.

dnephin replied:
Ok, even if this doesn't move into the updater creating a `getActiveNodes(ctx, client)` still goes a long way to making this more readable.

aaronlehmann replied:
Done

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+		tasks, err := client.TaskList(ctx, types.TaskListOptions{Filters: taskFilter})
+		if err != nil {
+			return err
+		}
```
Could this be a `getServiceTaskList(ctx, serviceID)` function called from the updater?

I believe the updater already has the necessary parameters to call this, and `taskFilter` could be defined much closer to where it's used (in the new `getServiceTaskList()` function.

It would also allow the `tasks` parameter to be removed from the interface definition.

aaronlehmann replied:
This would also require passing the `APIClient` into the updater. I have a slight preference for avoiding API calls inside the updater, and I don't think moving the service/task fetching there would yield much net simplification, but I can go ahead and do that if you want.

dnephin replied:
Ok, I understand wanting to avoid the API calls inside updater.

I'll ask for a smaller code change. `taskFilter` is defined above and not touched for 60 lines! Vertical separation makes it harder to read code. I think `taskFilter` should be moved to immediate above where it's used (line 122).

I would personally move both the `taskFilter` setup, and this api call into a `getUpToDateTasks(ctx, client, serviceID)` but I wont ask for that.

aaronlehmann replied:
Compromise: I've added a `getUpToDateTasks` closure near where `taskFilter` is defined. I don't think it make sense to repeatedly rebuild the filter within the loop.

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,378 @@
+				})
+			}
+			convergedAt = time.Time{}
+		}
```
I think this block could be a new function, something like:

```go
convergedAt := handleConvergence(converged, convergedAt, progressOut)

```
I'm sure there is a better name for this function, but nothing is coming to mind right now.

aaronlehmann replied:
The function would also need to take `monitor` as an argument.

IMHO splitting this out into a function wouldn't help with readability.

dnephin reviewed (commented):
Overall looks good. 

It looks a little light on tests. Some of this would be pretty difficult to test, but smaller functions might help. I noticed there are a few long functions. I've suggested some changes to reduce the size, and make them a bit easier to read.

aaronlehmann commented on cli/command/service/update.go:
```diff
@@ -196,7 +199,35 @@ func runUpdate(dockerCli *command.DockerCli, flags *pflag.FlagSet, serviceID str
+	if err == nil {
+		err = <-errChan
+	}
+	return err
```
Done

aluzzardi commented:
Small concern: `-i` / `--interactive` conflicts with `docker container create --interactive`.

`-I`? 

Perhaps `--progress` (but `-p` is a big no no).

`--wait`? Although `-w` is taken as well (workdir)

Anyone, suggestions?

sirlatrom commented:
`--foreground`, `--synchronous`, `--atomic`, even `--attach` (as antonym to
proposed `--detach`).

On 20 Mar 2017 21:44, "Andrea Luzzardi" <notifications@github.com> wrote:

> Small concern: -i / --interactive conflicts with docker container create
> --interactive.
>
> -I?
>
> Perhaps --progress (but -p is a big no no).
>
> --wait? Although -w is taken as well (workdir)
>
> Anyone, suggestions?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/docker/docker/pull/31144#issuecomment-287891518>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ-oUBqWQq5ZYU7Jb6STzeslRHZrlHZks5rnuUcgaJpZM4ME8oJ>
> .
>

sirlatrom commented:
`--foreground`, `--synchronous`, `--atomic`, even `--attach` (as antonym to
proposed `--detach`).

aluzzardi commented:
Thanks!

`--synchronous` / `-s` wouldn't conflict with anything docker container related.

(`-f` in foreground is typically reserved for `force` in docker, it's not really atomic, and there's a `docker attach` command which might be confusing for `--attach`)

aaronlehmann commented:
I'm not a fan of `--synchronous` (even though I used the word in the title of this PR. The term makes sense for programmers but not so much for non-programmers.

sirlatrom commented:
`--await-completion/-c`, `--quiescence/-q`...

sirlatrom commented:
`--return-result/-r`

sirlatrom commented:
Frankly,thee use case I need this for the most is scripting, not the interactive progress update (however neat :smiley:), similar to either `docker run` without `--detach`, or with `--detach` followed by either `docker wait $id` or `docker logs -f $id` depending on the situation. `--progress` or `--interactive` does not fit so nicely with that...

aaronlehmann commented:
The naming sure is a pain. `-i` / `--interactive` probably isn't right, not only because it might be confused with `docker run --interactive`, but also because progress bars aren't actually interactive. `--wait` doesn't feel quite right to me, and `-w` is already taken.

Some daemons have options like `-d` / `--dont-detach`. That's not really right here either because this is a client command, not a daemon.

How about `-P` / `--progress`, with a separate, mutually exclusive `--wait` option that waits but does not show progress. This feels most intuitive so far to me. The main downside is that it doesn't match the `--quiet` pattern used by most other commands (`--progress --quiet` seems contradictory - I don't think we should allow it). Note that rsync has a `-P` / `--progress` flag.

sirlatrom commented:
`-P` is `--publish-all` in `docker run`/`docker create`.

sirlatrom commented:
If naming consistency is important, the option should really be `--detach=false` with default value `true`. That would be most consistent with other docker commands, especially `docker create` and `docker run`, which are the commands most users will probably come from and expect likeness with.

dnephin commented on cli/command/service/progress/progress.go:
```diff
@@ -0,0 +1,398 @@
+)
+
+type progressUpdater interface {
+	update(ctx context.Context, service swarm.Service, tasks []swarm.Task, activeNodes map[string]swarm.Node, rollback bool) (bool, error)
```
I believe `ctx` is not used by either implementation, and can be removed from the interface.

dnephin commented:
I'm liking `--detach=false` / `-d=false`, and `-q --quiet` for the flags. I think that will eventually allow us to transition to defaulting to `-detach=false`. and the `-d` flag will work like `docker container run`.

aaronlehmann commented:
Switched to `--detach=false`. Also squashed commits.

aaronlehmann commented:
Also changed the progress bars to move backwards when rolling back (cc @aluzzardi).

dnephin reviewed (approved):
LGTM

aluzzardi commented:
:/

While I like the philosophy of `-d=false`, it's a PITA to type every time.

We didn't make this the default behavior and now it's a flag that needs to be turned off manually.

> I think that will eventually allow us to transition to defaulting to -detach=false

When? I don't see what's going to make it easier to transition to this in the future rather than now

dnephin commented:
> When? I don't see what's going to make it easier to transition to this in the future rather than now

In this release we add a `WARN: In a future release -d=false will become the default` for any `service create/update` that doesn't explicit set a `-d`

In either the next release or one after we change the default and remove the warning.

aaronlehmann commented:
> When? I don't see what's going to make it easier to transition to this in the future rather than now

I discussed this with @aluzzardi and @thaJeztah. It seems like the general policy would be to make a note in `docs/deprecated.md`, then flip the default in 3 stable releases, which would be about a year from now.

My sense is that a year is probably too long to wait. I don't think this extra time helps people prepare; rather it just lets the existing default get further entrenched. Intuitively waiting 1 or 2 stable releases feels right to me. Is it possible for us to make an exception to the general policy, since this is not really a deprecation (more of a change in behavior)?

thaJeztah commented:
Discussed in the maintainers meeting: We can keep it for 1 release, and based on feedback decide to keep it for 1 more release.

We can do a follow-up PR / issue to make it configurable in the CLI configuration file, but can decide on that one if we want it

aaronlehmann commented:
Thanks for discussing it. Do you want me to add a warning to the CLI when `-d` is not specified? What should the warning communicate?

dnephin commented:
My suggestion was something like this: "In a future release --detach=false will become the default"

aaronlehmann commented:
I've added warnings when the flag is not specified.

For example:

```
Since --detach=false was not specified, tasks will be updated in the background.In a future release, --detach=false will become the default.
```

PTAL

@aluzzardi: Are you okay with moving forward with this given the plan put forward here? https://github.com/docker/docker/pull/31144#issuecomment-290496884

thaJeztah commented:
I just realized that `--detach` / `-d` is also an option on `docker run`; if we want to reuse that flag in future to implement `--interactive` / `-i` and `--detach` options for docker services, similar to the way `docker run` works, we should consider using a different name for this flag to prevent it becoming ambiguous.

aaronlehmann commented:
The consensus so far seems to support `--detach` specifically because it's most consistent with `docker run`: https://github.com/docker/docker/pull/31144#issuecomment-288165223

I don't see a potential conflict between `docker run --detach` and `docker service create --detach`. It wouldn't really make sense for a service to remain "attached" to the terminal for the life of the service. There is not necessarily a single container to attach to; there may be many replicas. In the context of a service, `--detach` meaning to detach after convergence makes sense. At worst there's potential for confusion because the `run` flag and `service` flag won't behave the same way. But the other flag names proposed so far have their own downsides as well; I don't think there's a perfect answer.

I don't have strong feelings about the naming, but it's proving difficult to find a flag name that's acceptable to everyone. The discussion has already been going on for a few weeks.

dnephin commented:
@thaJeztah What differences do you see between this `-d` flag being added to service commands, and the `-d` flag on `run` ? I think they are the same idea, so we would use the same flag for consistency.

`-i` doesn't really relate to `-d` in any way that I can tell.

vieux commented:
I agree with @dnephin I think see how `-i` and `-d` are related. If you do not specify `-d` you're attached, whether or not you have `-i`

I'm 👍 on `--detach`

thaJeztah commented:
If there's no ambiguity, then all is fine (sorry, I meant to reply sooner). Just wanted to make sure we didn't overlook anything

thaJeztah reviewed (approved):
LGTM, it's really cool!

We can bike shed after merge if the error is too much "in your face" or if we have thoughts on a better approach for this (the config-file option that was discussed e.g.)

vieux reviewed (approved):
LGTM

aaronlehmann commented:
Thanks everyone!

nathanleclaire commented:
Is there a lil asciinema or GIF of this?

aaronlehmann commented:
@nathanleclaire: https://asciinema.org/a/9f9rgk0e7fczkxi5vfu7yofmw (slightly out of date, but mostly the same as what was merged)

marcosnils commented:
Hey guys, per @vieux request [Play With Docker](http://play-with-docker.com) has been updated with this patch. You can play with it there. Let me know in case you need me to change anything.

sirlatrom commented:
@aaronlehmann Great to see this merged 👍 Is there going to be a follow-up for `docker service scale`? I know that I can do `docker service update --detach=false --replicas=5`, but `docker service scale` allows scaling multiple services more than one service at a time, which is potentially a more complex operation to show than updating a single service.

Also, a minor detail: `--replicas 5`, `--name my_service`, etc. work, but `--detach false` does not; only `--detach=false`. Is this intentional?

thaJeztah commented:
> but `--detach false` does not

Could it be that (because `false` is a built-in / shell command), that the shell evaluates `false` if there's no `=` in between?

sirlatrom commented:
@theJeztah No, that would require me putting `false` inside backticks or a subshell. I haven't checked if the `--disable-content-trust` argument to e.g. `docker run` behaves the same, but I don't see why it should.

thaJeztah commented:
@sirlatrom interesting; could you open an issue for that, so that it can be looked into? It should be either fixed, or at least documented

sirlatrom commented:
@theJeztah I believe this is how the go flags package (or whatever is being used in the Docker CLI) works, so it's really just working as designed. The `--interactive/-i` and `--tty/-t` flags to `docker run` work the same way, and one can specify e.g. `--interactive=false` (which is the default in that case).

aaronlehmann commented:
It's because boolean flags can take an *optional* argument. `--name` requires an argument, so in `--name my_service`, `my_service` can be assumed to be the argument. But `--detach` is valid on its own, so `false` could be an entirely separate argument.

nathanleclaire commented:
@marcosnils Cool!

ozlatkov commented:
This looks great, thank you.

Just a quick one if anyone is aware - are there any plans to integrate this with

docker stack deploy

which in a nutshell creates/updates a collection of services?

Thanks again!

thaJeztah commented:
@ozlatkov see https://github.com/moby/moby/issues/32367

ozlatkov commented:
@thaJeztah  - Thank you.
